core/state_processor.go

```diff
@@ -179,6 +181,49 @@ func ApplyTransaction(config *params.ChainConfig, bc ChainContext, author *commo
 	return receipt, cxReceipt, gas, err
 }
 
+// ApplyStakingTransaction attempts to apply a staking transaction to the given state database
+// and uses the input parameters for its environment. It returns the receipt
+// for the staking transaction, gas used and an error if the transaction failed,
+// indicating the block was invalid.
+// staking transaction will use the code field in the account to store the staking information
+func ApplyStakingTransaction(
+	config *params.ChainConfig, bc ChainContext, author *common.Address, gp *GasPool, statedb *state.DB,
+	header *block.Header, tx *staking.StakingTransaction, usedGas *uint64, cfg vm.Config) (receipt *types.Receipt, gas uint64, err error) {
+
+	msg, err := StakingToMessage(tx, header.Number())
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Create a new context to be used in the EVM environment
+	context := NewEVMContext(msg, header, bc, author)
+
+	// Create a new environment which holds all relevant information
+	// about the transaction and calling mechanisms.
+	vmenv := vm.NewEVM(context, statedb, config, cfg)
+	// Apply the transaction to the current state (included in the env)
+	gas, err = ApplyStakingMessage(vmenv, msg, gp)
```

Comment:
Why there is no "failed" field to indicate failure. And if there is a failure, the receipt should include the failure reason.

Reply:
The failure indicates the code execution is failed but the process is no error. Here we use error to indicate any bad thing happen. It means we will either get error or always success (unlike the case of smart contract code exec)

Reply:
No, things like changing commssion rate that it's out of the MaxChangeRate will cause failure. And the user should be notified with a failure.